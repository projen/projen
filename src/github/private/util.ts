export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
