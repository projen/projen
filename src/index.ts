export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
