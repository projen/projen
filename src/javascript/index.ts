export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
