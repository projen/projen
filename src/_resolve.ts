import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
