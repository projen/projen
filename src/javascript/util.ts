import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
