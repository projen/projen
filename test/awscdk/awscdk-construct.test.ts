import { mkdirSync, writeFileSync } from "fs";
import { join } from "path";
import * as YAML from "yaml";
import { awscdk, LogLevel, Testing } from "../../src";
import {
  AwsCdkConstructLibrary,
  AwsCdkConstructLibraryOptions,
} from "../../src/awscdk";
import { NpmAccess } from "../../src/javascript";
import { mkdtemp, synthSnapshot } from "../util";

describe("constructs dependency selection", () => {
  test("user-selected", () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion: "1.100.0",
      constructsVersion: "3.1337.0-ultimate",
    });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot["package.json"]?.peerDependencies?.constructs).toBe(
      "^3.1337.0-ultimate"
    );
    expect(
      snapshot["package.json"]?.devDependencies?.constructs
    ).toBeUndefined();
    expect(snapshot["package.json"]?.dependencies?.constructs).toBeUndefined();
  });

  test("for cdk 1.x", () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: "1.112.0" });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot["package.json"]?.peerDependencies?.constructs).toMatch(
      /^\^3\./
    );
    expect(
      snapshot["package.json"]?.devDependencies?.constructs
    ).toBeUndefined();
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/assertions"]
    ).toStrictEqual("^1.112.0");
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/assert"]
    ).toStrictEqual("^1.112.0");
    expect(snapshot["package.json"]?.dependencies?.constructs).toBeUndefined();
  });

  // assertions library is only available since 1.111.0
  test("for cdk 1.x < 1.111.0", () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: "1.110.0" });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/assertions"]
    ).toBeUndefined();
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/assert"]
    ).toStrictEqual("^1.110.0");
  });

  test("for cdk 2.x", () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: "2.0.0-alpha.5" });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot["package.json"]?.peerDependencies?.constructs).toMatch(
      /^\^10./
    );
    expect(snapshot["package.json"]?.peerDependencies["aws-cdk-lib"]).toMatch(
      /^\^2./
    );
    expect(snapshot["package.json"]?.devDependencies?.constructs).toMatch(
      /^10./
    );
    expect(snapshot["package.json"]?.devDependencies["aws-cdk-lib"]).toMatch(
      /^2./
    );
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/assertions"]
    ).toBeUndefined();
    expect(snapshot["package.json"]?.dependencies?.constructs).toBeUndefined();
  });

  test("for cdk 2.x, throws if incorrect constructsVersion provided", () => {
    expect(
      () =>
        new TestProject({
          cdkVersion: "2.0.0-alpha.5",
          constructsVersion: "3.2.27",
        })
    ).toThrow(/CDK 2.x requires constructs 10.x/);
  });

  test("for cdk 2.x, throws if cdkDependencies provided", () => {
    expect(
      () =>
        new TestProject({
          cdkVersion: "2.0.0-alpha.5",
          cdkDependencies: ["@aws-cdk/aws-lambda"],
        })
    ).toThrow(
      /cdkDependencies is not used for CDK 2.x. Use "peerDeps" or "deps" instead/
    );
  });

  test("for cdk 2.x, throws if cdkTestDependencies provided", () => {
    expect(
      () =>
        new TestProject({
          cdkVersion: "2.0.0-alpha.5",
          cdkTestDependencies: ["@aws-cdk/aws-lambda"],
        })
    ).toThrow(
      /cdkTestDependencies is not used for CDK 2.x. Use "devDeps" or "testDeps" instead/
    );
  });

  test("for cdk 2.x, throws if cdkDependenciesAsDeps provided", () => {
    expect(
      () =>
        new TestProject({
          cdkVersion: "2.0.0-alpha.5",
          cdkDependenciesAsDeps: true,
        })
    ).toThrow(/cdkDependenciesAsDeps is not used for CDK 2.x/);
  });

  test("for cdk 2.x, throws if cdkAssert provided", () => {
    expect(
      () =>
        new TestProject({
          cdkVersion: "2.0.0-alpha.5",
          cdkAssert: true,
        })
    ).toThrow(
      /cdkAssert is not used for CDK 2.x. Use the assertions library that is provided in aws-cdk-lib/
    );
  });

  test("for cdk 3.x (does not exist yet)", () => {
    // GIVEN
    expect(() => new TestProject({ cdkVersion: "3.1337.42" })).toThrow(
      /Unsupported AWS CDK major version 3\.x/
    );
  });
});

describe("lambda functions", () => {
  test("are auto-discovered by default", () => {
    // GIVEN
    const outdir = mkdtemp();
    mkdirSync(join(outdir, "src"));
    writeFileSync(join(outdir, "src", "my.lambda.ts"), "// dummy");

    const project = new TestProject({
      cdkVersion: "1.100.0",
      libdir: "liblib",
      outdir: outdir,
      bundlerOptions: {
        assetsDir: "resources",
      },
      lambdaOptions: {
        runtime: awscdk.LambdaRuntime.NODEJS_22_X,
        bundlingOptions: {
          externals: ["foo", "bar"],
          sourcemap: true,
        },
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot["src/my-function.ts"]).not.toBeUndefined();
    expect(
      snapshot[".projen/tasks.json"].tasks["bundle:my.lambda"].steps
    ).toStrictEqual([
      {
        exec: 'esbuild --bundle src/my.lambda.ts --target="node22" --platform="node" --outfile="resources/my.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:foo --external:bar --sourcemap',
      },
    ]);
  });

  test("auto-discover can be disabled", () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion: "1.100.0",
      lambdaAutoDiscover: false,
    });

    // WHEN
    mkdirSync(join(project.outdir, project.srcdir));
    writeFileSync(
      join(project.outdir, project.srcdir, "my.lambda.ts"),
      "// dummy"
    );

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot["src/my-function.ts"]).toBeUndefined();
    expect(snapshot[".projen/tasks.json"].tasks["bundle:my"]).toBeUndefined();
  });
});

describe("node version in workflow", () => {
  it("does setup default version", () => {
    const project = new TestProject({ cdkVersion: "2.12.0" });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build).toMatchSnapshot();
    expect(buildWorkflow.jobs.build.steps).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          uses: expect.stringContaining("actions/setup-node"),
          with: {
            "node-version": "lts/*",
          },
        }),
      ])
    );
  });

  it("does use minNodeVersion", () => {
    const project = new TestProject({
      cdkVersion: "2.12.0",
      minNodeVersion: "18.0.0",
    });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build).toMatchSnapshot();
    expect(buildWorkflow.jobs.build.steps).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          uses: expect.stringContaining("actions/setup-node"),
          with: {
            "node-version": "18.0.0",
          },
        }),
      ])
    );
  });

  it("does setup a custom version", () => {
    const project = new TestProject({
      cdkVersion: "2.12.0",
      minNodeVersion: "16.0.0",
      workflowNodeVersion: "20.17.0",
    });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build).toMatchSnapshot();
    expect(buildWorkflow.jobs.build.steps).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          uses: expect.stringContaining("actions/setup-node"),
          with: {
            "node-version": "20.17.0",
          },
        }),
      ])
    );
  });
});

describe("workflow container image", () => {
  it("does not use an image by default for cdk v1", () => {
    const project = new TestProject({ cdkVersion: "1.100.0" });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build).not.toHaveProperty("container");
  });

  it("does not use an image by default for cdk v2", () => {
    const project = new TestProject({ cdkVersion: "2.12.0" });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build).not.toHaveProperty("container");
  });

  it("uses the user-defined image if specified", () => {
    const project = new TestProject({
      cdkVersion: "2.12.0",
      workflowContainerImage: "my-custom-image",
    });
    const snapshot = synthSnapshot(project);
    const buildWorkflow = YAML.parse(snapshot[".github/workflows/build.yml"]);
    expect(buildWorkflow.jobs.build.container.image).toStrictEqual(
      "my-custom-image"
    );
  });
});

describe("integ-runner", () => {
  test('adds "integ-runner" to devDependencies', () => {
    const project = new TestProject({
      cdkVersion: "2.12.0",
      experimentalIntegRunner: true,
    });

    const snapshot = synthSnapshot(project);

    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/integ-runner"]
    ).toStrictEqual("latest");
    expect(
      snapshot["package.json"]?.devDependencies["@aws-cdk/integ-tests-alpha"]
    ).toStrictEqual("latest");
    expect(project.tasks.tryFind("integ")?.steps).toEqual([
      { exec: "integ-runner $@ --language typescript", receiveArgs: true },
    ]);
    expect(project.tasks.tryFind("integ:update")?.steps).toEqual([
      {
        exec: "integ-runner $@ --language typescript --update-on-failed",
        receiveArgs: true,
      },
    ]);
    expect(project.testTask.steps).toEqual(
      expect.arrayContaining([{ spawn: "integ" }])
    );
  });
});

it("warns the user if they add CDK v1 dependencies to a CDK v2 project", () => {
  // GIVEN
  console.error = jest.fn();
  const project = new TestProject({
    cdkVersion: "2.12.0",
    deps: [
      "@aws-cdk/core",
      "@aws-cdk/aws-s3",
      "@aws-cdk/cfnspec",
      "@aws-cdk/aws-apigatewayv2-alpha@2.25.0-alpha.0",
    ],
    logging: {
      level: LogLevel.VERBOSE,
    },
  });

  // WHEN
  Testing.synth(project);

  // THEN
  expect(console.error).toHaveBeenCalledWith(
    expect.stringContaining(
      `WARNING: Found CDK v1 deps in your project, even though your "cdkVersion" is 2.x: [@aws-cdk/aws-s3, @aws-cdk/core].`
    )
  );
});

const defaultOptions = {
  author: "Nobody",
  authorAddress: "nobody@nowhere.com",
  clobber: false,
  defaultReleaseBranch: "main",
  jest: false,
  name: "test-project",
  npmAccess: NpmAccess.PUBLIC,
  repositoryUrl: "https://github.com/projen/projen.git",
} as const;

class TestProject extends AwsCdkConstructLibrary {
  constructor(
    options: Omit<AwsCdkConstructLibraryOptions, keyof typeof defaultOptions>
  ) {
    super({
      ...defaultOptions,
      ...options,
    });
  }
}
