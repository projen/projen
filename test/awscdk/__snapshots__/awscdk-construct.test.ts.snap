// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node version in workflow does setup a custom version 1`] = `
{
  "env": {
    "CI": "true",
  },
  "outputs": {
    "self_mutation_happened": "\${{ steps.self_mutation.outputs.self_mutation_happened }}",
  },
  "permissions": {
    "contents": "write",
  },
  "runs-on": "ubuntu-latest",
  "steps": [
    {
      "name": "Checkout",
      "uses": "actions/checkout@v4",
      "with": {
        "ref": "\${{ github.event.pull_request.head.ref }}",
        "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
      },
    },
    {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v4",
      "with": {
        "node-version": "20.17.0",
      },
    },
    {
      "name": "Install dependencies",
      "run": "yarn install --check-files",
    },
    {
      "name": "build",
      "run": "npx projen build",
    },
    {
      "id": "self_mutation",
      "name": "Find mutations",
      "run": "git add .
git diff --staged --patch --exit-code > repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT",
      "shell": "bash",
      "working-directory": "./",
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Upload patch",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "repo.patch",
        "overwrite": true,
        "path": "repo.patch",
      },
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Fail build on mutation",
      "run": "echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
cat repo.patch
exit 1",
    },
    {
      "continue-on-error": true,
      "name": "Backup artifact permissions",
      "run": "cd dist && getfacl -R . > permissions-backup.acl",
    },
    {
      "name": "Upload artifact",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "build-artifact",
        "overwrite": true,
        "path": "dist",
      },
    },
  ],
}
`;

exports[`node version in workflow does setup default version 1`] = `
{
  "env": {
    "CI": "true",
  },
  "outputs": {
    "self_mutation_happened": "\${{ steps.self_mutation.outputs.self_mutation_happened }}",
  },
  "permissions": {
    "contents": "write",
  },
  "runs-on": "ubuntu-latest",
  "steps": [
    {
      "name": "Checkout",
      "uses": "actions/checkout@v4",
      "with": {
        "ref": "\${{ github.event.pull_request.head.ref }}",
        "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
      },
    },
    {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v4",
      "with": {
        "node-version": "lts/*",
      },
    },
    {
      "name": "Install dependencies",
      "run": "yarn install --check-files",
    },
    {
      "name": "build",
      "run": "npx projen build",
    },
    {
      "id": "self_mutation",
      "name": "Find mutations",
      "run": "git add .
git diff --staged --patch --exit-code > repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT",
      "shell": "bash",
      "working-directory": "./",
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Upload patch",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "repo.patch",
        "overwrite": true,
        "path": "repo.patch",
      },
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Fail build on mutation",
      "run": "echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
cat repo.patch
exit 1",
    },
    {
      "continue-on-error": true,
      "name": "Backup artifact permissions",
      "run": "cd dist && getfacl -R . > permissions-backup.acl",
    },
    {
      "name": "Upload artifact",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "build-artifact",
        "overwrite": true,
        "path": "dist",
      },
    },
  ],
}
`;

exports[`node version in workflow does use minNodeVersion 1`] = `
{
  "env": {
    "CI": "true",
  },
  "outputs": {
    "self_mutation_happened": "\${{ steps.self_mutation.outputs.self_mutation_happened }}",
  },
  "permissions": {
    "contents": "write",
  },
  "runs-on": "ubuntu-latest",
  "steps": [
    {
      "name": "Checkout",
      "uses": "actions/checkout@v4",
      "with": {
        "ref": "\${{ github.event.pull_request.head.ref }}",
        "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
      },
    },
    {
      "name": "Setup Node.js",
      "uses": "actions/setup-node@v4",
      "with": {
        "node-version": "18.0.0",
      },
    },
    {
      "name": "Install dependencies",
      "run": "yarn install --check-files",
    },
    {
      "name": "build",
      "run": "npx projen build",
    },
    {
      "id": "self_mutation",
      "name": "Find mutations",
      "run": "git add .
git diff --staged --patch --exit-code > repo.patch || echo "self_mutation_happened=true" >> $GITHUB_OUTPUT",
      "shell": "bash",
      "working-directory": "./",
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Upload patch",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "repo.patch",
        "overwrite": true,
        "path": "repo.patch",
      },
    },
    {
      "if": "steps.self_mutation.outputs.self_mutation_happened",
      "name": "Fail build on mutation",
      "run": "echo "::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch."
cat repo.patch
exit 1",
    },
    {
      "continue-on-error": true,
      "name": "Backup artifact permissions",
      "run": "cd dist && getfacl -R . > permissions-backup.acl",
    },
    {
      "name": "Upload artifact",
      "uses": "actions/upload-artifact@v4.4.0",
      "with": {
        "name": "build-artifact",
        "overwrite": true,
        "path": "dist",
      },
    },
  ],
}
`;
